ext/drx/drx.h

```diff
@@ -435,6 +435,19 @@ DR_EXPORT
 size_t
 drx_buf_get_buffer_size(void *drcontext, drx_buf_t *buf);
 
+
+DR_EXPORT
+/**
+ * Pads a basic block with a label for routines which rely on inserting instrumentation
+ * after every instruction. Note that users of this routine must act on the previous
+ * instruction before the is-app check afterwards, because the label is not marked as an
+ * app instruction.
+ *
+ * \returns whether terminator was introduced.
```

Comment:
For the time being, the work is mostly done. I have a curious assert failure on Windows 64-bit, deep in the `interp.c` code, but otherwise here is what I have done so far:
1. In `interp.c` we currently assert that if we found a cti or a syscall, it is the last instruction in its basic block. Easy fix.
2. In drreg, we will lazily restore all apps at the very end of the basic block. I've modified this so that we only unreserve registers at the last non-label instruction of a basic block.